C# 9 reference update: lambda discard parameters

pkulikov opened the pull request:
Contributes to #20577

pkulikov committed:
C# 9 reference update: lambda discard parameters

pkulikov commented on docs/csharp/language-reference/operators/lambda-expressions.md:
```diff
@@ -44,44 +44,57 @@ When you use method-based syntax to call the <xref:System.Linq.Enumerable.Select
+
+You cannot use statement lambdas to create expression trees.
+
+## Input parameters of a lambda expression
```
That information was inside the section about expression lambdas, while it's true for all lambdas. So, I've extracted the section.

pkulikov commented on docs/csharp/language-reference/operators/lambda-expressions.md:
```diff
@@ -44,44 +44,57 @@ When you use method-based syntax to call the <xref:System.Linq.Enumerable.Select
 ## Expression lambdas
 
-A lambda expression with an expression on the right side of the `=>` operator is called an *expression lambda*. Expression lambdas are used extensively in the construction of [expression trees](../../programming-guide/concepts/expression-trees/index.md). An expression lambda returns the result of the expression and takes the following basic form:
+A lambda expression with an expression on the right side of the `=>` operator is called an *expression lambda*. An expression lambda returns the result of the expression and takes the following basic form:
```
Removed that sentence as it breaks the flow of the paragraph; moreover, that information is mentioned earlier in the article.

pkulikov commented on docs/csharp/language-reference/operators/lambda-expressions.md:
```diff
@@ -216,10 +229,8 @@ The following rules apply to variable scope in lambda expressions:
-## Featured book chapter
+For more information about lambda discard parameters, see the [feature proposal note](~/_csharplang/proposals/csharp-9.0/lambda-discard-parameters.md)
 
-[Delegates, Events, and Lambda Expressions](/previous-versions/visualstudio/visual-studio-2008/ff518994(v=orm.10)) in [C# 3.0 Cookbook, Third Edition: More than 250 solutions for C# 3.0 programmers](/previous-versions/visualstudio/visual-studio-2008/ff518995(v=orm.10))  
```
This article has been updated considerably during the last years; I guess, it's very far from that original chapter now (which is for C# 3)

BillWagner replied:
Yes, that's true.

Editorial tidbit: When I last updated Effective C# and More Effective C# (for versions 6 & 7, respectively), one of my reviewers made the point of commenting whenever I missed a reference to a "new" feature from C# 3 (the previous updates.)

pkulikov reviewed (commented):
@BillWagner I've left a number of clarifying comments for some of the changes

IEvangelist commented on docs/csharp/language-reference/operators/lambda-expressions.md:
```diff
@@ -44,44 +44,57 @@ When you use method-based syntax to call the <xref:System.Linq.Enumerable.Select
+Beginning with C# 9.0, you can use [discards](../../discards.md) to specify two or more input parameters of a lambda expression that aren't used in the expression:
 
-## Statement lambdas
+[!code-csharp[discard parameters](snippets/lambda-expressions/ExpressionAndStatementLambdas.cs#Discards)]
```
```suggestion
:::code language="csharp" source="snippets/lambda-expressions/ExpressionAndStatementLambdas.cs" id="SnippetDiscards":::
```

IEvangelist commented on docs/csharp/language-reference/operators/lambda-expressions.md:
```diff
@@ -44,44 +44,57 @@ When you use method-based syntax to call the <xref:System.Linq.Enumerable.Select
+[!code-csharp[discard parameters](snippets/lambda-expressions/ExpressionAndStatementLambdas.cs#Discards)]
 
-A statement lambda resembles an expression lambda except that the statement(s) is enclosed in braces:
+That feature may be useful when you use a lambda expression to [provide an event handler](../../programming-guide/events/how-to-subscribe-to-and-unsubscribe-from-events.md).
```
```suggestion
Lambda discard parameters may be useful when you use a lambda expression to [provide an event handler](../../programming-guide/events/how-to-subscribe-to-and-unsubscribe-from-events.md).
```

IEvangelist reviewed (commented):
These are awesome changes, well done. I had a few bits of feedback and I'll let @BillWagner handle it from here. As always, thank you @pkulikov - great job!

pkulikov commented on docs/csharp/language-reference/operators/lambda-expressions.md:
```diff
@@ -44,44 +44,57 @@ When you use method-based syntax to call the <xref:System.Linq.Enumerable.Select
-Specify zero input parameters with empty parentheses:  
+## Statement lambdas
+
+A statement lambda resembles an expression lambda except that the statement(s) is enclosed in braces:
```
@IEvangelist  Should it be "its statements" instead of "their"?

IEvangelist replied:
I'd be good with that, although I believe that using their in this context is fine too.

pkulikov committed:
Update docs/csharp/language-reference/operators/delegate-operator.md

Co-authored-by: David Pine <david.pine@microsoft.com>

pkulikov committed:
Update docs/csharp/language-reference/operators/lambda-expressions.md

Co-authored-by: David Pine <david.pine@microsoft.com>

pkulikov committed:
Update docs/csharp/language-reference/operators/lambda-expressions.md

Co-authored-by: David Pine <david.pine@microsoft.com>

pkulikov committed:
Update docs/csharp/language-reference/operators/lambda-expressions.md

Co-authored-by: David Pine <david.pine@microsoft.com>

pkulikov commented:
@IEvangelist Thank you for review and suggestions.

pkulikov committed:
Update lambda-expressions.md

BillWagner reviewed (commented):
This is great @pkulikov 


I have only one general comment beyond what @IEvangelist mentioned (which have already been addressed).

I'd like to see a link to these changes from the C# 9 what's new article. That helps readers find more detailed information.

pkulikov committed:
Addressed feedback

pkulikov commented:
@BillWagner thank you. I've updated the What's new article with the links (also added the links that should have been added in the previous similar PRs).

BillWagner reviewed (approved):
You are awesome!   🎆 Thank you.

I'll approve this and :shipit: now.
